server/src/main/java/keywhiz/service/providers/ClientAuthFactory.java

```diff
@@ -89,18 +89,22 @@ public Client provide(ContainerRequest request) {
   }
 
   private static class MyAuthenticator implements Authenticator<String, Client> {
-    private final ClientDAO clientDAO;
+    private final ClientDAO clientDAOReadWrite;
+    private final ClientDAO clientDAOReadOnly;
 
-    private MyAuthenticator(ClientDAO clientDAO) {
-      this.clientDAO = clientDAO;
+    private MyAuthenticator(
+        ClientDAO clientDAOReadWrite,
+        ClientDAO clientDAOReadOnly) {
+      this.clientDAOReadWrite = clientDAOReadWrite;
+      this.clientDAOReadOnly = clientDAOReadOnly;
     }
 
     @Override public Optional<Client> authenticate(String name)
         throws AuthenticationException {
-      Optional<Client> optionalClient = clientDAO.getClient(name);
+      Optional<Client> optionalClient = clientDAOReadOnly.getClient(name);
       if (optionalClient.isPresent()) {
         Client client = optionalClient.get();
-        clientDAO.sawClient(client);
+        clientDAOReadWrite.sawClient(client);
```

Comment:
We could do this write conditionally if the last seen was >1 week or something like that.